### DISCUSSION TITLE
route not found error

### DISCUSSION THREAD

**Discussion by simbwa** [coupergateway/couper#716] [Q&A]
Hi folks.

Getting a "route not found error" from avenga/couper when I put an API gateway powered by avenga/couper behind Traefik.

Essentially, Traefik is the primary web proxy and avenger/couper acts as middleware to transform GET requests with query params to JSON to be posted to an internal microservice that accepts only JSON POST requests.

```hcl
server {
  api {
  endpoint "/prod/rx/9000/mo" {
    proxy {
        backend {
            origin = env.BACKEND_ORIGIN
            path = env.BACKEND_PATH
        }
    }
    request "request" {
      json_body = {
	    id = request.id
        msgType = "mo"
        shortCode = "9K"
        srcId = "XX1"
        srcTypeId = "aggregator"
        msisdn = request.query.phone_no[0]
        msg = request.query.message[0]
        b64encoded = base64_encode(request.query.message[0])
        tstamp = unixtime()
        srcReq = {
            URL = request.url
            Headers = request.headers
        }
      }
    }
    response {
    	json_body = backend_responses.request.json_body
    }
  }
  }
}
```

Am testing out avenger/couper for the first time and am still learning & experimenting with it. 
Please point me in the right direction on how to fix this issue.

**Comment by filex** [comment]
Hi @simbwa , can you post the full error message and ideally the corresponding line from Couper's log output?

**Comment by malud** [comment]
Hi @simbwa ,
The `request` block seems to be missing a target like the `url` or `backend` attribute. If you just want to transform a request to post specific content from the client request to an origin you just need the `request` and optionally a `response` block. You can remove the `proxy` block. The `proxy` block would forward the client request as is to the configured backend.

This works with `curl -v http://localhost:8080/prod/rx/9000/mo\?message=asdf`

```hcl
server {
  api {
    endpoint "/prod/rx/9000/mo" {
      request {
        method = "POST"
        json_body = {
          id = request.id
          msgType = "mo"
          shortCode = "9K"
          srcId = "XX1"
          srcTypeId = "aggregator"
          msisdn = request.query.phone_no[0]
          msg = request.query.message[0]
          b64encoded = base64_encode(default(request.query.message[0], "empty"))
          tstamp = unixtime()
          srcReq = {
            URL = request.url
            Headers = request.headers
          }
        }

        backend = "my_service"
      }

      response {
        json_body = backend_responses.default.json_body.data
      }
    }
  }
}

definitions {
  backend "my_service" {
    origin = "https://httpbin.org"
    path = "/post"
  }
}
```

You could also following this example which could match your case: https://github.com/avenga/couper-examples/tree/master/custom-requests

**Path not found**: If you still see a not found error from Couper, change the `endpoint "/**"` path to a wildcard one and watch the logs which path Traefik uses at the end.

**Reply by simbwa** [reply]
Thanks @malud 

Here is the twist 

This works

`curl -v http://localhost:8080/prod/rx/9000/mo\?message=asdf`

But this does not work - access behind Traefik - returns a 404 error

`curl -v https://event.mydomain.tld/prod/rx/9000/mo?phone_number=0789&message=test123`

 Here are the logs

```
{
    "build": "8ea8f09",
    "client_ip": "10.0.1.85",
    "handler": "file",
    "level": "error",
    "message": "route not found error",
    "method": "GET",
    "port": "80",
    "request": {
        "headers": {
            "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.2 Safari/605.1.15"
        },
        "host": "event.mydomain.tld",
        "method": "GET",
        "origin": "event.mydomain.tld:80",
        "path": "/prod/rx/9000/mo?phone_number=0789\u0026message=test123",
        "proto": "http",
        "status": 404,
        "tls": false
    },
    "response": {
        "bytes": 8016,
        "headers": {
            "content-encoding": "gzip",
            "content-type": "text/html"
        }
    },
    "status": 404,
    "timestamp": "2023-02-12T10:25:55Z",
    "timings": {
        "total": 0.765
    },
    "type": "couper_access",
    "uid": "cfkbtcp8md2c73dfpl90",
    "url": "http://event.mydomain.tld:80/prod/rx/9000/mo?phone_number=0789\u0026message=test123",
    "version": "1.12.0"
}
```

**Reply by simbwa** [reply]
@malud I updated my couper.hcl to match what you shared. But still no luck behind Traefik.

@filex I have posted the full error message from the logs for your review.

Thanks

**Reply by simbwa** [reply]
Extra context.

Deployed  in a docker swarm cluster with the Traefik as the proxy for all exposed services. Not sure if that detail is of significance or not.

**Reply by malud** [reply]
Hi, did you changed the default port for Couper or making use of the `hosts` attribute?
You may want to share your swarm config if its possible? Can't reproduce this at the moment.
thanks.

**Reply by simbwa** [reply]
@malud Sorry for my late response. In the end, I swapped out couper with Nginx and all worked well. That is why I didn't revert sooner. 
Thanks all for your help.
